### PR TITLE
[refactor] Add resources to signatures

### PR DESCRIPTION
**Pull request by croyzor**
Resource requirements for a dataflow node can be represented by a pair of the resources requirements of the inputs and the resource requirements of the outputs (see the very bottom of the diff).

Inside of a dataflow node, the input requirements are uniformly associated with the wires coming out of the source (Input) node, and the output requirements are associated with the sinks in the Output node

In this PR:
* Add input & output resources to signatures
* Change function interfaces to deal with signatures when possible, so that resources aren't forgotten about
* Include resource requirements in Input/Output nodes, so that their signatures can be computed

### CODE REVIEW

File: src/ops/validate.rs [hunk 1]
```diff
@@ -513,9 +514,11 @@ mod test {
 
         let input_node = OpType::Dataflow(DataflowOp::Input {
             types: in_types.clone(),
+            resources: ResourceSet::new(),
```
Comment (ss2165) [inline src/ops/validate.rs:517]: on main `Input` and `Output` are now standalone structs, so you can define constructors for them to make this easier.
I suggest 
```rust 
fn new(types: TypeRow) -> Self;
fn with_resources(self, resources: ResourceSet) -> Self;
```
 So you can do `ops::Input::new(type_row![QB]).with_resources(...)`

Comment (croyzor) [reply src/ops/validate.rs]: Done!

File: src/builder/tail_loop.rs [hunk 2]
```diff
@@ -20,14 +20,13 @@ impl<B: HugrMutRef> TailLoopBuilder<B> {
     pub(super) fn create_with_io(
         base: B,
         loop_node: Node,
-        tail_loop: &ops::TailLoop,
+        tail_loop_sig: &ops::TailLoop,
```
Comment (ss2165) [inline src/builder/tail_loop.rs:23]: this shouldn't have been renamed

File: src/ops/dataflow.rs [hunk 3]
```diff
@@ -26,25 +27,64 @@ pub(super) trait DataflowOpTrait {
     }
 }
 
+pub(crate) trait IOTrait {
```
Comment (ss2165) [inline src/ops/dataflow.rs:30]: I think this can be `pub`, generally useful for construction

**Review by ss2165 — APPROVED** [review]

**Pull request merged**